letmeind/src/protocol.rs

```diff
@@ -121,8 +121,28 @@ impl<'a, C: ConnectionOps> Protocol<'a, C> {
         self.resource_id = None;
         self.auth_state = AuthState::NotAuth;
 
-        // Receive the initial knock message.
-        let knock = self.recv_msg(Operation::Knock).await?;
+        // Receive the initial message (Knock or Close).
+        let initial_msg = match timeout(self.conf.control_timeout(), self.conn.recv_msg())
```

Comment:
Please change `self.recv_msg` instead of open-coding the timeout and expected operation check here.
We could use a reference to a slice of expected operations as a parameter to `self.recv_msg`.